lib/oauth/multi.go

```diff
@@ -0,0 +1,182 @@
+			var loginState LoginState
+			if _, err := enc.Decode(context.Background(), []byte(fString), &loginState); err != nil {
+				errs = append(errs, fmt.Errorf("%v, was %s expected %s, err %v", errStateNotType, reflect.TypeOf(state), reflect.TypeOf(&MultiOAuthState{}), err))
+			}
+			if s, ok := loginState.State.(MultiOAuthState); ok {
+				return &s, nil
+			}
+		}
+		return nil, multierror.New(errs)
+	} else if fs, ok := state.(MultiOAuthState); ok {
+		return &fs, nil
+	}
+	return nil, fmt.Errorf("%v, was %s expected %s", errStateNotType, reflect.TypeOf(state), reflect.TypeOf(&MultiOAuthState{}))
+}
+
+// decodeState will check if the state alignment is present. If it isnt it will return a
+func (mo *MultiOauth) decodeState(r *http.Request) (*MultiOAuthState, error) {
+	query := r.URL.Query()
+	state := query.Get("state")
+	// No state means fresh flow
+	if state == "" {
+		if mo.OptAuth == nil {
+			return &MultiOAuthState{}, nil
+		}
+		return &MultiOAuthState{}, nil
+	}
+	return mo.decodeRaw(state)
+}
+
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+// authenticator will return the specified oauth.Authenticator if MultiOAuthState.CurrentFlow is set.
+// if it is not, it will spit our a random optional flow or the last, required flow. It also will set MultiOAuthState.CurrentFlow.
+func (mo *MultiOauth) authenticator(state *MultiOAuthState) (string, *Authenticator) {
+	if state.CurrentFlow != "" {
+		return state.CurrentFlow, mo.OptAuth[state.CurrentFlow]
+	}
+	// find the first authen not redeemed, if they are all redeemed return required
+	for k, v := range mo.OptAuth {
+		if !contains(state.CompletedFlows, k) {
+			state.CurrentFlow = k
+			return k, v
+		}
+	}
+	state.CurrentFlow = requiredKey
+	return requiredKey, mo.RequiredAuth
+}
+
+func (mo *MultiOauth) PerformAuth(w http.ResponseWriter, r *http.Request, mods ...kcookie.Modifier) (AuthData, bool, error) {
+	flowState, err := mo.decodeState(r)
+	if err != nil {
+		return AuthData{}, false, err
+	}
+	key, a := mo.authenticator(flowState)
+	data, _, err := a.PerformAuth(w, r, mods...)
+	if err != nil {
+		return AuthData{}, false, err
+	}
+	// This is the terminal condition.
+	if key == requiredKey {
+		data.PrimaryIdentity = data.Creds.Identity
+		data.Identities = flowState.OptIdentities
+		data.State = flowState.Extra
+		return data, true, err
+	}
+	flowState.OptIdentities = append(flowState.OptIdentities, data.Creds.Identity)
+	flowState.CompletedFlows = append(flowState.CompletedFlows, key)
+	flowState.CurrentFlow = ""
+	data.State = flowState
+	return data, false, err
+}
+
+func (mo *MultiOauth) PerformLogin(w http.ResponseWriter, r *http.Request, lm ...LoginModifier) error {
+	options := LoginModifiers(lm).Apply(&LoginOptions{})
+	// if the passed in state does not match what we want, pack in the previous state to unpack later.
+	if _, ok := options.State.(*MultiOAuthState); !ok {
```

Comment:
This issue is that decodeRaw handles options.State which is an interface{} and url query state which is a string.

So It's used in two places that have two different types. I prefer just having the decode raw because a lot of wire deserialization structs have ti and it's pretty intuitive.

Reply:
as per chat over VC, the if here is decoding the state already, so no reason to call decodeRaw anymore. At that point, decodeRaw is only ever called with a string as parameter, and can be simplified.